### DISCUSSION TITLE
How does ca_id really work?

### DISCUSSION THREAD

**Discussion by hgl** [strongswan/strongswan#1604] [Q&A] [answered]
I just found out that ca_id was added in 5.8.2, and I'm trying to understand how it works.

It seems if `ca_id = CN=abc` is set on the responder, then it would accept a client's certificate chain if any (sub) CA's CN is abc?

This doesn't sound right, because the client could totally generate a chain itself with that CN and gets accept.

I think I must've misunderstood how it works. Would be grateful if anyone could shed some light.

**Comment by tobiasbrunner** [comment, marked as answer]
The complete certificate chain has to be trusted to begin with. `ca_id` only serves as an additional constraint. That's similar to `cacerts`, but that  requires the certificates locally when the config is loaded, which might not be the case for some intermediate CA certificates. And the certificates in `cacerts` also have to binary match, `ca_id` simplifies that.

**Edit:** You can find an example in the [ikev2-multi-ca/rw-hash-and-url](https://strongswan.org/testing/testresults/ikev2-multi-ca/rw-hash-and-url/) test scenario.

**Reply by hgl** [reply]
Thanks for the quick reply.

It starts to make sense. So am I correct that
1. ca_id only applies to intermediate CAs, but not the root CA?
2. Root CA still has to be locally loaded in order for the whole chain to be trusted?
3. ~~ca_id allows an initiator to omit sending intermediate CAs, and picks the chain if any subject value matches the value?~~ (this question doesn’t make sense, if intermediate CAs are not sent, there is no chain. I was drunk)

Thanks for the example, but `swanctl --list-certs` on moon lists all intermediate and root CAs, which seems to mean they are all locally loaded. So I can’t come up with answers from it to my previous questions.

**Reply by tobiasbrunner** [reply]
> 1. ca_id only applies to intermediate CAs, but not the root CA?

No, it can be used as constraint against root CAs (e.g. to select different connections based on the CA).

> 2. Root CA still has to be locally loaded in order for the whole chain to be trusted?

Yes.

> 3\. ~ca_id allows an initiator to omit sending intermediate CAs, and picks the chain if any subject value matches the value?~ (this question doesn’t make sense, if intermediate CAs are not sent, there is no chain. I was drunk)

Yes, the intermediate CA certificates have to be either loaded locally or sent by the peer (in the test scenario they are sent in their hash-and-URL form, so they are technically not sent but fetched by the gateway).

> Thanks for the example, but `swanctl --list-certs` on moon lists all intermediate and root CAs, which seems to mean they are all locally loaded.

No, that command runs at the end of the test case. At that point the intermediate CA certificates were fetched and are now cached and therefore listed (and available for later authentications, i.e. they don't have to be fetched again). You can see in moon's log which certificates are loaded when the IKE daemon is started.

**Reply by hgl** [reply]
Thanks for the detailed reply. It makes perfect sense now.
